python/lsst/ctrl/bps/wms/htcondor/lssthtc.py

```diff
@@ -0,0 +1,836 @@
+        Initial data.
+    """
+    def __init__(self, valid_keys, init_data=()):
+        self.valid_keys = valid_keys
+        self.data = {}
+        self.update(init_data)
+
+    def __getitem__(self, key):
+        return self.data[key]
+
+    def __delitem__(self, key):
+        del self.data[key]
+
+    def __setitem__(self, key, value):
+        """Stores key,value in internal dict only if key is valid
+
+        Parameters
+        ----------
+        key : `str`
+        value : `object`
+        """
+        if key not in self.valid_keys:
+            raise KeyError(f"Invalid key {key}")
+        self.data[key] = value
+
+    def __iter__(self):
+        return self.data.__iter__()
+
+    def __len__(self):
+        return len(self.data)
+
+    def __str__(self):
+        return str(self.data)
+
+
+def htc_escape(value):
+    """Escape characters in given string based upon HTCondor syntax.
+
+    Parameter
+    ----------
+    value : `str`
+        String that needs to have characters escaped.
+
+    Returns
+    new_value : `str`
+        Given string with characters escaped.
+    """
+    return value.replace("\\", "\\\\").replace('"', '\\"').replace("'", "''")
+
+
+def htc_write_attribs(stream, attrs):
+    """Write job attributes in HTCondor format to writeable stream.
+
+    Parameters
+    ----------
+    stream : `TextIOBase`
+        Output text stream (typically an open file)
+    attrs : `dict`
+        HTCondor job attributes (dictionary of attribute key, value)
+    """
+    for key, value in attrs.items():
+        stream.write(f'+{key} = "{htc_escape(value)}"\n')
+
+
+def htc_write_condor_file(filename, job_name, job, job_attrs):
+    """Main function to write an HTCondor submit file.
+
+    Parameters
+    ----------
+    filename : `str`
+        Filename for the HTCondor submit file
+    job_name : `str`
+        Job name to use in submit file
+    job : `RestrictedDict`
+        Submit script information.
+    job_attrs : `dict`
+        Job attributes.
+    """
+    os.makedirs(os.path.dirname(filename), exist_ok=True)
+    with open(filename, "w") as fh:
+        for key, value in job.items():
+            if key in HTC_QUOTE_KEYS:
+                fh.write(f'{key}="{htc_escape(value)}"\n')
+            else:
+                fh.write(f"{key}={value}\n")
+        for key in ["output", "error", "log"]:
+            if key not in job:
+                filename = f"{job_name}.$(Cluster).${key[:3]}"
+                fh.write(f"{key}={filename}\n")
+
+        if job_attrs is not None:
+            htc_write_attribs(fh, job_attrs)
+        fh.write("queue\n")
+
+
+def htc_version():
+    """Return the version given by the HTCondor API.
+
+    Returns
+    -------
+    version : `str`
+        HTCondor version as easily comparable string.
+
+    Raises
+    -------
+    RuntimeError
+        Raised if fail to parse htcondor API string.
+    """
+    # $CondorVersion: 8.8.6 Nov 13 2019 BuildID: 489199 PackageID: 8.8.6-1 $
```

Comment:
I _think_ this is an example of a string that would match the regex below. I think it would be worth saying so in a comment.